File: projects/json/json.hs
module JSON where

data JValue = JString String
            | JNumber Double
            | JObject [(String, JValue)]
            | JArray [JValue]
            | JTrue
            | JFalse
            | JNull

instance Show JValue where
   show (JString a) = show a
   show (JNumber a) = show a

   show (JArray []) = "[]"
   show (JArray a)  = "[" ++ showArr a ++ "]"
      where showArr (x:[]) = show x
            showArr (x:xs) = show x ++ ", " ++ showArr xs

   show (JObject []) = "{}"
   show (JObject a)  = "{" ++ showObj a ++ "}"
      where showObj ((k,v):[]) = show k ++ ": " ++ show v
            showObj ((k,v):xs) = show k ++ ": " ++ show v ++ ", " ++ showObj xs

   show (JTrue)  = "true"
   show (JFalse) = "false"
   show (JNull)  = "null"

--instance Read JValue where
--  readsPrec _ _ = [(JString "test", "test")]

testObj :: JValue
testObj = JObject [("first_name", JString "oliver"), ("last_name", JString "stollmann")]

testObj2 :: JValue
testObj2 = JObject [("name", JObject [("first_name", JString "Oliver"), ("last_name", JString "Stollmann")]), ("age", JNumber 22), ("info", JArray [JString "Zinistr. 7", JString "8004 Zuerich", JString "Schweiz"]), ("rand", JString "babybaby")]

File: projects/json/json.py
import sys

class JValue(object):
   def __init__(self,value=None):
       self._value = value

class JString(JValue):
  def __repr__(self):
    return "\"" + self._value + "\""

class JNumber(JValue):
  def __repr__(self):
    return str(self._value)

class JBool(JValue):
  def __repr__(self):
    if self._value == True:
      return "true"
    elif self._value == False:
      return "false"

class JNull(JValue):
  def __repr__(self):
    return "null"

class JObject(JValue):
   def __init__(self):
       self.__objdict = {}

   def __getitem__(self, key):
       return self.__objdict[key]

   def __setitem__(self, key, jpair):
       self.__objdict[jpair.key()] = jpair.value()

   def __repr__(self):
       return 'JObject'

class JArray(JValue):
   def __init__(self):
       self._value = []

   def append(self,item):
       self._value.append(item)

   def __getitem__(self,key):
       return self._value[key]

   def __setitem__(self,key,value):
       self._value[key] = value

   def __repr__(self):
       return '[' + ', '.join(map(str, self.__arrlist)) + ']'


class JPair(JValue):
   def __init__(self,key=None,value=None):
       # key: JString
       self.key = key
       # value: JValue
       self.value = value

   def key(self):
       return self.key

   def value(self):
       return self.value

   def __repr__(self):
       return '%s: %s' % (self.key,self.value)


if __name__ == '__main__':
   arr = JArray()
   print(arr)
   obj = JObject()
   print(obj)
   pair = JPair("first_name", "last_name")
   print(pair)
